### ISSUE TITLE
Add readManyAsync with timeout method to Ringbuffer api

### ISSUE THREAD

**Issue by ahmetmircik** [hazelcast/hazelcast#17610]
__Reason:__
To prevent possible OOM as in [this issue](https://github.com/hazelcast/hazelcast/issues/17061#issue-633502934):

__Possible signature:__
```
readManyAsync(long startSequence, int minCount, int maxCount,                                                      
     IFunction<E, Boolean> filter, longtimeout, TimeUnit timeUnit)
```

**Comment by mmedenjak** [comment]
@ahmetmircik I don't think we need to add this in 4.1 as it's not a regression. As far as I understand, the operation is queued, with some preallocated array and instantiated objects, and never gets executed.

Cancelling the operation when `future.get()` times out doesn't seem right so the right approach is, as you say, adding a new method.

But for 4.1 I'm thinking, can we at least avoid (delay) allocating the large array and some other costly objects when the operation is simply parked?

**Comment by ahmetmircik** [comment]
@mmedenjak yes, setting it to  4.1 is only for rising awareness. I also think, it can be moved to another milestone.

Regarding lazy allocation, i will investigate for it.

**Comment by mmedenjak** [comment]
Thanks!
